### PR TITLE
4upstream

### LINKED ISSUE #41: no method for setting pgain/ggain
I know that the policy for this lib is to not expose all registers to keep the footprint small. But I think it really needs methods for setting the proximity-gain (PGAIN) and gesture-gain (GGAIN). The default is 1x and 4x respectively. With  this setting, proximity detection kicks in at a distance below an inch (at least for my sensor), which makes the driver almost useless.

The lib already has a setter/getter for the color-gain (AGAIN), and I would provide a patch with similar methods for the PGAIN/GGAIN settings if you also think this is a useful addition.

### PR DESCRIPTION
**Pull request by bablokb**
add getter/setter for proximity-gain (PGAIN)  and gesture-gain (GGAIN).

### CODE REVIEW

**Comment by ladyada** [comment]
hiya! thanks so much for submitting a PR! we can review & merge PRs once they have passed continuous integration (CI). that means that code is 'linted' - that means it is formatted and passes basic structure tests, so that we maintain the same text formatting for all new code
if your code isnt passing, check the CI output (click on the red X next to the PR to scroll through the log and find where the error is

here is a tutorial on pylint and black: https://learn.adafruit.com/improve-your-code-with-pylint

and overall how to contribute PRs to circuitpython: https://learn.adafruit.com/contribute-to-circuitpython-with-git-and-github

once you get that green checkmark that indicates CI has passed, please comment reply to this post so we know its time for another review (we may not get notified on CI pass and miss that its time to look!)

**Comment by bablokb** [comment]
Hm, the log does not tell me why the check failed.

**Comment by ladyada** [comment]
you need to run `black`, that part failed

```black....................................................................Failed```

**Comment by bablokb** [comment]
Hi, checks from CI are now successful

File: adafruit_apds9960/apds9960.py [hunk 1]
```diff
@@ -195,7 +201,7 @@ def __init__(
             self._write8(_APDS9960_GCONF4, 0)
             self._write8(_APDS9960_GPULSE, 0)
             self._write8(_APDS9960_ATIME, 255)
-            self._write8(_APDS9960_CONTROL, 1)
+            self._write8(_APDS9960_CONTROL, 0)
```
Comment (ladyada) [inline adafruit_apds9960/apds9960.py:204]: can you revert this or explain it please!

**Review by ladyada — CHANGES_REQUESTED** [review]

File: adafruit_apds9960/apds9960.py (hunk 1 above)
Comment (bablokb) [reply adafruit_apds9960/apds9960.py:204]: There are two reasons for this change: first, the code claims to set registers to power-on defaults, but according to the datasheet the power-on default for register CONTROL is 0x00. Bits 1:0 of this register are the AGAIN setting, and the existing code (setter-method color_gain) also documents that the power-on default value for AGAIN is 0.

The driver defaults for AGAIN is 1 and is set a few lines down.

**Comment by ladyada** [comment]
@caternuson wanna check it?

**Comment by caternuson** [comment]
Looks good and generally works:
```python
Adafruit CircuitPython 7.1.1 on 2022-01-14; Adafruit ItsyBitsy M4 Express with samd51g19
>>> import board
>>> from adafruit_apds9960.apds9960 import APDS9960
>>> apds = APDS9960(board.I2C())
>>> apds.proximity_gain
0
>>> apds.proximity_gain = 3
>>> apds.proximity_gain
3
>>> apds.gesture_gain
2
>>> apds.gesture_gain = 0
>>> apds.gesture_gain
0
>>> 
```

@bablokb What do you think about adding a simple dictionary (or similar) to map the gain values, like `8x` to the register values, like `3` and use that behind the scenes. That way users could think in terms of the actual gain values, for example:
```python
# set gain to 8x 
apds.proximity_gain = 8
```
It looks like both prox gain and gesture gain have the same general mapping. So could use one for both.

And throw a `ValueError` exception for any invalid value.
```python
# this should throw an exception
apds.proximity_gain = 100
# and this, since the is no 3x gain
apds.proximity_gain = 3
```

**Comment by bablokb** [comment]
With a dictionary PGAIN and GGAIN would have a different logic than AGAIN (color_gain), unless you want to break existing code. What about adding constants `PGAIN_1X` and so on? We could add these constants also for color-gain without breaking things. But new code would be more readable. 

Checking values and throwing ValueErrors is currently not consistent within the library. I will be happy to add that, but there is a footprint tradeoff. So this is more of a policy decision.

File: adafruit_apds9960/apds9960.py (hunk 1 above)
Comment (caternuson) [reply adafruit_apds9960/apds9960.py:204]: yep, datasheet for ref:
image

**Comment by caternuson** [comment]
Thanks. Good point about `color_gain` (AGAIN). I think being consistent is better at this point. My suggestion can be deferred to some later PR. You're also right that it'd be breaking. So would require some consideration.

@ladyada lgtm, including the `_APDS9960_CONTROL` change you flagged.

**Review by caternuson — APPROVED** [review]

**Comment by ladyada** [comment]
thanks :)

**Pull request merged**